File: csharp/895.maximum-frequency-stack.cs
/*
 * @lc app=leetcode id=895 lang=csharp
 *
 * [895] Maximum Frequency Stack
 */

using System.Max;
using System.Collections.Generic;

// @lc code=start
public class FreqStack {

    private int Max = 0;
    private IDictionary<int, int> Frequency;
    private IDictionary<int, IList<int>> Stacks;

    public FreqStack() {
        Frequency = new Dictionary<int, int>();
        Stacks = new Dictionary<int, IList<int>>();
    }
    
    public void Push(int x) {
        if (!Frequency.TryGetValue(x, out int count)) {
            count = 0;
        }
        ++count;
        if (!Stacks.TryGetValue(count, out var stack)) {
            stack = new List<int>();
            Stacks[count] = stack;
        }
        Frequency[x] = count;
        stack.Add(x);
        Max = Math.Max(Max, count);
    }
    
    public int Pop() {
        var stack = Stacks[Max];
        int count = stack.Count, ret = stack[count - 1];
        stack.RemoveAt(count - 1);
        if (count == 1) {
            --Max;
        }
        --Frequency[ret];
        return ret;
    }
}

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 */
// @lc code=end

File: java/895.maximum-frequency-stack.java
/*
 * @lc app=leetcode id=895 lang=java
 *
 * [895] Maximum Frequency Stack
 */

// @lc code=start
class FreqStack {

    private Map<Integer, Integer> freqs;
    private Map<Integer, Stack<Integer>> elements;
    private int max;

    public FreqStack() {
        freqs = new HashMap<Integer, Integer>();
        elements = new HashMap<Integer, Stack<Integer>>();
        max = 0;
    }
    
    public void push(int x) {
        int freq = freqs.getOrDefault(x, 0) + 1;
        freqs.put(x, freq);
        if (!elements.containsKey(freq)) {
            elements.put(freq, new Stack<Integer>());
        }
        elements.get(freq).push(x);
        max = Math.max(freq, max);
    }
    
    public int pop() {
        Stack<Integer> curr = elements.get(max);
        int ret = curr.pop();
        freqs.put(ret, max - 1);
        if (curr.size() == 0) {
            --max;
        }
        return ret;
    }
}

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 */
// @lc code=end

